app/controllers/corrections/CorrectionCountryController.scala

```diff
@@ -32,41 +35,67 @@ import scala.concurrent.{ExecutionContext, Future}
 
 
   protected val controllerComponents: MessagesControllerComponents = cc
 
-  def onPageLoad(mode: Mode, period: Period, index: Index): Action[AnyContent] = cc.authAndGetDataAndCorrectionToggle(period) {
+  def onPageLoad(mode: Mode, period: Period, periodIndex: Index, countryIndex: Index): Action[AnyContent] = cc.authAndGetDataAndCorrectionToggle(period) {
     implicit request =>
-      val form = formProvider(index,
+      val form = formProvider(periodIndex,
         request.userAnswers
-          .get(CorrectionCountryPage).toSeq)
+          .get(CorrectionCountryPage(periodIndex, countryIndex)).toSeq)
 
-      val preparedForm = request.userAnswers.get(CorrectionCountryPage) match {
+      val preparedForm = request.userAnswers.get(CorrectionCountryPage(periodIndex, countryIndex)) match {
         case None => form
         case Some(value) => form.fill(value)
       }
 
-      Ok(view(preparedForm, mode, period, index))
+      request.userAnswers.get(CorrectionReturnPeriodPage(periodIndex)) match {
+        case Some(correctionPeriod) => Ok(view(preparedForm, mode, period, periodIndex, correctionPeriod, countryIndex))
+        case None => Redirect(controllers.routes.JourneyRecoveryController.onPageLoad().url)
+      }
   }
 
-  def onSubmit(mode: Mode, period: Period, index: Index): Action[AnyContent] = cc.authAndGetDataAndCorrectionToggle(period).async {
+  def onSubmit(mode: Mode, period: Period, periodIndex: Index, countryIndex: Index): Action[AnyContent] = cc.authAndGetDataAndCorrectionToggle(period).async {
     implicit request =>
-      val form = formProvider(index,
+      val form = formProvider(periodIndex,
         request.userAnswers
-          .get(CorrectionCountryPage).toSeq)
+          .get(CorrectionCountryPage(periodIndex, countryIndex)).toSeq)
 
       form.bindFromRequest().fold(
         formWithErrors =>
-          Future.successful(BadRequest(view(formWithErrors, mode, period, index))),
+          request.userAnswers.get(CorrectionReturnPeriodPage(periodIndex)) match {
+            case Some(correctionPeriod) => Future.successful(BadRequest(view(formWithErrors, mode, period, periodIndex, correctionPeriod, countryIndex)))
+            case None => Future.successful(Redirect(controllers.routes.JourneyRecoveryController.onPageLoad().url))
+          },
 
         value =>
-          for {
-            updatedAnswers <- Future.fromTry(request.userAnswers.set(CorrectionCountryPage, value))
-            _              <- cc.sessionRepository.set(updatedAnswers)
-          } yield Redirect(CorrectionCountryPage.navigate(mode, updatedAnswers))
+          request.userAnswers.get(CorrectionReturnPeriodPage(periodIndex)) match {
+            case Some(correctionPeriod) =>
+              for {
+                          updatedAnswers <- Future.fromTry(request.userAnswers.set(CorrectionCountryPage(periodIndex, countryIndex), value))
+                          vatReturnResult <- vatReturnConnector.get(correctionPeriod)
```

Comment:
Can we swap again the connector call and repo call so userAnswers update regardless

Reply:
does it need to go inside the yield or just be after the repo call?

Reply:
done